### ISSUE TITLE
ReadableStream should be an async iterable

### ISSUE THREAD

**Issue by mariusGundersen** [whatwg/streams#778]
(I'm adding this mostly as a tracking bug after I [asked about it in the wrong place](https://github.com/whatwg/fetch/issues/586)) 

NodeJS is [looking into supporting async iterables](https://github.com/tc39/proposal-async-iteration/issues/74) as a way to stream data, and it would be great if fetch (or the readable stream part of fetch) supported the same interface. This would make easier to move code between nodejs and the browser.

**Comment by jakearchibald** [comment]
https://jakearchibald.com/2017/async-iterators-and-generators/#making-streams-iterate - article on this, including basic implementation.

**Comment by mariusGundersen** [comment]
Excellent! I really look forward to this!

**Comment by guest271314** [comment]
@jakearchibald Getting two errors at jsbin 

`Uncaught SyntaxError: Unexpected token *` 

when both `async` and `*` are used, and

`Uncaught SyntaxError: Unexpected reserved word`

at `await` before `for..of` loop at Chromium 60 with `--js-flags=--harmony-async-iteration` flag set.

Can you provide link documentations relevant to `--js-flags=--harmony-async-iteration` flag, which does not appear to be listed at [List of Chromium Command Line Switches](https://peter.sh/experiments/chromium-command-line-switches/)?

`await` can be used both before `iterable` at `(variable of iterable)` and within `for..of` loop `statement`. If expected result is for iteration to wait for current `Promise` value following `await`. For example


```
// wait for each `iterable` before reaching `statement`
(async() => {
  for (let prop of [await new Promise(resolve => 
      setTimeout(resolve, Math.floor(Math.random() * 10000), 1)), await new Promise(resolve => 
      setTimeout(resolve, Math.floor(Math.random() * 5000), 2))]
      ) {
    console.log(prop);
  };
 })();
```

```
// use generator function for `iterable`
(async(n) => {
  function* gen() {
    while(++n < 5) yield new Promise(resolve => setTimeout(resolve, 1000, {value:n, done:false}))
  }

  for (let prop of gen()) {
    console.log(await prop)
  }
})(0);
```

```
// pass an array of functions to be called within `statement`,
// next `iterable` awaits fulfillment of `Promise` following `await` 
const fn = async(res, ...props) => {
  for (let prop of props) {
    res.push(await prop());
    console.log(res);
  };
  return res
}

const arr = [
  () =>
    new Promise((resolve, reject) => 
      setTimeout(() =>
        resolve(0)
      , Math.floor(Math.random() * 2000))
    ),
  () => 
    new Promise((resolve, reject) => 
      setTimeout(() => 
        resolve(1) 
      , Math.floor(Math.random() * 2000))
    )
  ];

fn([], ...arr)
.then(data => console.log("done:", data))
.catch(err => console.error(err));
```



Adjusted pattern at linked jsbin https://jsbin.com/zusexeripu/edit?js,console to not combine `async` with `*` at function declaration for number example

```
function* asyncRandomNumbers() {
  // This is a web service that returns a random number
  const url = 'https://www.random.org/decimal-fractions/?num=1&dec=10&col=1&format=plain&rnd=new';
  while(true) yield fetch(url).then(response => response.text()).then(text => Number(text));  
}
async function example() {
  for (const p of asyncRandomNumbers()) {
    const n = await p;
    
    if (n > 0.95) {
      console.log("done", n);
      break;
    } else {
        console.log(n);
    }
  }
}

example();
```

To iterate a `ReadableStream` we can use several different patterns, depending on expected flow control and result

```
async function readStream() {
  let n = 0;
  const rs = new ReadableStream({
    pull(c) {
      if (n++ < 5) {
        c.enqueue(n)
      } else {
        c.close()
      }
    }
  });
  const reader = rs.getReader();
  const res = [];
  async function processStream() {
    return await reader.read();
  }

  function* gen() {
    while(true) yield processStream()
  }
  
  for (let stream of gen()) {
    const {value, done} = await stream; 
    if (done) break;
    res.push(value);
    console.log(value, done);
  }

  return {res, rs, reader, complete: await reader.closed.then(() => true)};
  
};

```

**Comment by domenic** [comment]
There are two approaches to this:

- Have `[Symbol.asyncIterable]` directly return the default reader object, and add aliases to reader: `next()` aliases `read()`; `return()` aliases `releaseLock()`. And I guess make `releaseLock()` return an empty object?
- Have `[Symbol.asyncIterable]` return a small wrapper object, of the sort in Jake's blog post.

After realizing we'd have to make `releaseLock()` return an empty object, I'm leaning toward the latter. But, does anyone think the former would be better?

**Comment by domenic** [comment]
Further questions that have come up as I contemplate this all a bit more seriously:

- Should the async iterator's return() (which is called, remember, when you break out of a for-await-of loop via return/break/throw) cancel the reader, or just release the lock? I.e., if you break out of a loop, should we assume you'll never want to continue looping from the current point and we should clean up for you? On the one hand, that seems presumptuous. On the other hand, it's kind of annoying to make people manually acquire a new reader and call `cancel()` if they break out, especially if that break out is due to an unexpected throw.
- If you iterator to the end, should we release the lock for you? This is somewhat of an edge case, and just governs what happens if someone tries to acquire a new iterator or lock after iterating to the end.

I'm leaning toward auto-canceling and auto-releasing, because otherwise, cleanup is quite annoying:

```js
try {
  for await (const chunk of rs) {
    await somethingThatMightReject(chunk);
  }
} finally {
  try {
    // Might throw if the reader is still locked because we finished successfully.
    await rs.cancel();
  } catch {}
}
```

It's starting to feel a bit magic though....

**Comment by mariusGundersen** [comment]
[Here](https://github.com/es-git/es-git/blob/master/packages/http-transport/ts/utils/streamToAsyncIterator.ts) is the implementation that I currently use to convert a ReadableStream to an async iterator, and it works well enough for what I need from it. I don't think I ever cancel iterating the stream though, maybe only when throwing an exception in a loop. 

I also agree that it should be auto cancelled and auto closed, as (async) iterators most often are single consumer. I think the api should be optimized for single consumer and require a small wrapper for cases where it should not automatically close. That seems like the most common scenario, and it doesn't make it impossible to get the multi consumer scenario to work.

**Comment by guest271314** [comment]
Auto-cancelling and auto-releasing should suffice for a base implementation. 

FWIW [this code](https://gist.github.com/guest271314/74d3a2aa765330163eecc46e16a52acd) began with an attempt to place a possibly infinite `Promise` chain within a [`while` expression](https://stackoverflow.com/q/47171474). A `Map` object is used for the purpose of dynamically adding elements to be iterated (from perspective here, the default implementation of trying to pass a dynamic value to `next().then()` is difficult to arrage; we would first need define the value of `yield` as a variable and use an addition line or two for `if..else` within the generator function). A flag is used which should stop iteration immediately. The iteration can also be stopped by returning or throwing. All errors should reach and be handled at the last chained `.catch()` after propagating through the previous `.catch()`s, though that can be adjusted to continue processing if a `Promise` is rejected within process.

**Comment by jakearchibald** [comment]
Wrapper object seems like a clear winner.

I didn't like the sound of auto-cancel, but given @domenic's code example it sounds like the better thing to do.

We could have:

```js
for await (const chunk of stream.iterator({ preventClose: true })) {
  // …
}
```

Then `stream[Symbol.asyncIterator]` could alias `stream.iterator`.

**Comment by mcollina** [comment]
@jakearchibald @domenic I would like Node.js `Readable` to be as close as possible to whatwg `Readable` when used with `for await`, i.e. `stream.iterator(opts)`. I think it would help defining a best practice on how iterators should be used. Also, it would improve code portability.
Is this API spec'd yet? Could I be involved in the process?

One thing that is not obvious, is if we should close the stream when there is a `break`.

**Comment by jakearchibald** [comment]
@mcollina see the above comments for API discussion so far. Is there somewhere we can read up on the justifications and tradeoffs for the pattern node is shipping?

**Comment by domenic** [comment]
`break` invokes `return()`, which per the above design will cancel the stream. I'm not sure it's "obvious" that it should cancel the stream, but it's probably better for the common case, and @jakearchibald had a good idea for how we could allow an escape hatch for the uncommon case.

**Comment by mcollina** [comment]
@jakearchibald I added AsyncIterators based on what I thought it might make sense for Node.js Streams, we also did a couple of implementations and this turned out to be more performant*. It's an experimental feature that is shipping for the first time in 2 weeks (it will print a warning and notify the users). We can definitely change any part of what I did in [implementation](https://github.com/nodejs/node/blob/master/lib/internal/streams/async_iterator.js) before it gets out of experimental (ideally before Node 10 goes LTS in October, but even later if things are in flux).

> break invokes return(), which per the above design will cancel the stream.

I guess my implementation of `return`  should be correct then, thanks!

* It will need to be benchmarked again in the future, V8 is shipping a lot of optimizations for promises, generators and AsyncIterators.

**Comment by domenic** [comment]
@devsnek has generously volunteered to help with the spec and tests here, and I'm very excited. Summing up some IRC discussion:

- I think we should have auto-release. There's no reason for any consumer to forever hold a lock and not have any way of giving it back, even if the stream is definitely closed. At least in tests we often check whether something's errored or closed by doing `.getReader().closed.then(...)`; we shouldn't break that, I think.
- I agree with auto-cancel as the default return() behavior. I like @jakearchibald's idea of `.iterator({ preventCancel: true })` as a way of customizing. (Jake said "preventClose", but "preventCancel" is more accurate, unless I'm missing something.)

**Comment by benjamingr** [comment]
> I think we should have auto-release.

In a large code base using our own stream->to->async iterator transition we did this and are pretty happy with that as a default.

>  At least in tests we often check whether something's errored or closed by doing .getReader().closed.then(...); we shouldn't break that, I think.

Definitely agree from our perspective.

> I agree with auto-cancel as the default return() behavior.

We also do this and find it quite useful. It is also easy to opt out of even without providing `.iterator({ preventCancel: true })` by either:

 - Wrapping the stream with another stream for the purpose of that iteration and not forwarding cancellation.
 - Wrapping the async iterator with another async iterator and not forwarding the `.return` there.

That said, I am not objecting to a `.iterator({ preventCancel: true })` just saying that we haven't had to use it in our own code.


------

I've also been using async iterators with Node streams which have been working well - though I admit a lot less than I'd like to - I have been trying to (as in emailing relevant parties) solicit feedback but haven't been able to get it. We hope to solicit feedback through a Node.js foundation survey which we hope to send in about a month.

**Comment by domenic** [comment]
Given the precedent that is somewhat being established in https://github.com/WICG/kv-storage/issues/6, I am wondering whether we should rename `.iterator()` to `.values()`, hmm.

**Comment by ricea** [comment]
I prefer `.iterator()` for ReadableStream. I think people are likely to see `.values()` and expect it to be a shortcut for slurping the whole stream. `.iterator()` is more specific about the functionality it provides.

**Comment by MattiasBuelens** [comment]
> I am wondering whether we should rename `.iterator()` to `.values()`, hmm.

> I prefer `.iterator()` for ReadableStream.

Uhm, at the moment it's called `.getIterator()` instead of `.iterator()`. If you want to rename it, let me know in #980. 😉

**Comment by ricea** [comment]
I forgot to file implementer bugs yesterday.

Here they are:
Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=929585
Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1525852
WebKit: https://bugs.webkit.org/show_bug.cgi?id=194379

**Comment by devilblacknick** [comment]
nicknickson2530@gmail.com
hi i am nk creak my girlfriend whatsapp plz how to creak help me .....!!!!!!

**Comment by liudonghua123** [comment]
I have some code like this. But vscode show some errors (`Type 'Promise<ReadableStreamReadResult<Uint8Array>>' must have a '[Symbol.asyncIterator]()' method that returns an async iterator.`) around `reader.read()`.

```js
    const response = await fetch(`${download_url_prefix}${node_executable_filename}`)
    const contentLength = +(response.headers.get('Content-Length') ?? 0);
    const reader = response.body?.getReader();
    if(!reader) {
      throw new Error(`Failed to get reader from response body`);
    }
    let receivedLength = 0;
    let chunks = []; 
    for await (let {done, value} of reader.read()) {
      chunks.push(value);
      receivedLength += value.length;
      console.log(`Received ${receivedLength} of ${contentLength}`)
    }
```

I have to rewrite the above `for await` to `while`.

```js
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      chunks.push(value);
      received_length += value.length;
      console.log(`Received ${receivedLength} of ${contentLength}`)
    }

```

**Comment by MattiasBuelens** [comment]
@liudonghua123 You shouldn't use `getReader()` or `reader.read()` in this case, instead you should `for await` over the `ReadableStream` itself:
```javascript
const readable = response.body;
if (!readable) {
  throw new Error('Failed to get response body');
}
let receivedLength = 0;
let chunks = []; 
for await (const value of readable) {
  chunks.push(value);
  receivedLength += value.length;
  console.log(`Received ${receivedLength} of ${contentLength}`)
}
```
[MDN has more examples on async-iterating a stream.](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#async_iteration_of_a_stream_using_for_await...of) 😉

**Comment by liudonghua123** [comment]
@MattiasBuelens Thanks, I see now. :smile:

**Comment by himself65** [comment]
I'm just conscious of why there's no type checking for ReadableStream. I got `any` when use AsyncIterator
<img width="531" alt="image" src="https://github.com/whatwg/streams/assets/14026360/9531095b-4db4-41b6-b790-42cd100dc657">

**Comment by MattiasBuelens** [comment]
@himself65 Your question is about the TypeScript type definitions, so you should look at https://github.com/microsoft/TypeScript/ and/or https://github.com/microsoft/TypeScript-DOM-lib-generator/.

That said, it's already been fixed in https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1701, so in a future version of TypeScript your code snippet should get proper type checking. (I don't know if this will already be available in version 5.5, I'm not with the TypeScript team.)
